### ISSUE TITLE
Relation#first does not GC

### ISSUE THREAD

**Issue by jrafanie** [rails/rails#21921]
After performing `Post.all.first` in the template report, the Post instance is not being garbage collected after it's out of scope.  Note, `Post.all.last` does GC.  Only the initial call to the `first` method appears to be affected.  If you call `first` fifty times afterwards, there is only the initial instance in `ObjectSpace`.

See the gist here:  https://gist.github.com/jrafanie/1a6a58ac00c0996f1ab2

This also occurs in 4.2.4: https://gist.github.com/jrafanie/c69a7bbd0a66bebbeda3

cc @matthewd @fryguy @dmetzger57

**Comment by jrafanie** [comment]
cc @tenderlove

**Comment by matthewd** [comment]
Fixed in 1b6fcae948b162efbd17f08ca563ff05742bf01b; backported to 4-2-stable as 599c31cf7d617955ea727358bf9878b365252f51

**Comment by mikeantonelli** [comment]
@jrafanie @matthewd This change breaks implementations utilizing `PerThreadRegistry` where the `singleton_class` does not have a defined delegate, no?

Updating to activesupport to 4.2.5 from 4.2.4 started failing my tests:

```
NoMethodError: undefined method `delegate' for #<Class:MyClass>
```
